rfc/0000-batch-job-groups.rst

```diff
@@ -0,0 +1,1179 @@
+performant! Instead, we create the job group based on jobs that have
+already been created at that point in time (filter on existing jobs
+rather than as a matching rule that is executed on each new job that
+is created). The implementation for this operation is to take a query
+filter string / job group definition and then find all matching jobs
+for that filter condition and assign them to the new job group. The
+creation operation will return a job group ID that can be used for
+subsequent polling and cancellation operations. A big concern with
+this approach is its O(n_jobs) and will be slow for large batches and
+it's likely the request will timeout before Batch can process the
+request. This use case necessitates the need for longer running
+idempotent async operations that the user can poll for completion of
+(for example, creating disks in GCP). A poor man's implementation for
+this operation is to have the client list jobs matching the query filter
+and then the client explicitly creates the new job group specifying the
+listed jobs.
+
+For the QoB use case, we know upfront which group we want to assign
+jobs to. Therefore, we will only implement the first interface for
+creating a job group and save the later interfaces for future work.
+
+
+~~~~~~~~~~~~~~~~~~
+Getting the Status
+~~~~~~~~~~~~~~~~~~
+
+Getting the status of a job group is a single HTTP request that
+executes an O(1) database query to do a small aggregation on the table
+that keeps track of the number of jobs in each state and the billing
+tables. The user must know the job group ID corresponding to the group
+or the server needs to have a mechanism for translating a job group
+"name" into an ID to query for.
+
+
+~~~~~~~~~~~~
+Cancellation
+~~~~~~~~~~~~
+
+Cancelling the job group is a single HTTP request and an O(1) database
+insert operation. The job group ID is inserted into a table that
+tracks which job groups have been cancelled.
+
+The autoscaler and scheduler avoid trying to spin up resources for
+jobs in job groups that have been cancelled, but the individual job
+has not been cleaned up yet by ignoring any jobs that are in cancelled
+batches or job groups (identical to the current behavior). For an
+accurate fair share computation, the modified
+``user_inst_coll_resources`` table keeps track of the number of ready
+jobs, running jobs, etc. per user, per instance collection, and now
+**per batch**. When a batch has been cancelled or a job group is
+actively being cancelled, then those rows of the table pertaining to
+the specific batch are skipped. This design is a rework of the current
+cancellable resources tables. Because we don't need to track the
+cancellable states of every job group, we can have job groups that
+don't follow a tree like structure and still be able to cancel them
+quickly and not have any performance regressions or incorrect fair
+share computations that affect other user's resource allocations and
+cluster efficiency.
```

Comment:
Why are we adding a per batch thing if the problem is job groups? How does this relate to `batch_inst_coll_cancellable_resources`?